### DISCUSSION TITLE
Word-level timestamps?

### DISCUSSION THREAD

**Discussion by eudoxos** [openai/whisper#332] [Q&A]
I was trying a simple
```py
import whisper
model=whisper.load_model("large")
result=model.transcribe("p_trim3.wav")
```
to see if I can locate timestamps for individual words/tokens in the result but I don't see them in the output. Is it possible to get this from the model?

I am reading these lines:

https://github.com/openai/whisper/blob/d18e9ea5dd2ca57c697e8e55f9e654f06ede25d0/whisper/tokenizer.py#L150-L151

https://github.com/openai/whisper/blob/d18e9ea5dd2ca57c697e8e55f9e654f06ede25d0/whisper/tokenizer.py#L191-L192

so in this example (one segment from a test audio), the numbers marked in red should be timestamps (in samples)? There is 4 words in the segment text, but only 3 parts in-between the 4 timestamps. Any help/hint will be much appreciated.

![image](https://user-images.githubusercontent.com/1029876/196029067-3c782bf6-6df0-427b-8932-53bf5ae20f97.png)

**Comment by Arlen22** [comment]
https://github.com/openai/whisper/blob/d18e9ea5dd2ca57c697e8e55f9e654f06ede25d0/whisper/decoding.py#L651

I think one of the lines here should have what you are looking for.

**Comment by jianfch** [comment]
this was discussed in #3

**Comment by jongwook** [comment]
I've made a demo of obtaining word-level timestamps using the cross-attention patterns in [the multilingual ASR notebook](https://github.com/openai/whisper/blob/main/notebooks/Multilingual_ASR.ipynb)

![image](https://user-images.githubusercontent.com/266841/210467619-4384e3d0-522b-4e84-8c64-60e6422eed26.png)

**Comment by Jeronymous** [comment]
I improved the solution of @jongwook (thanks to him for the very useful notebook!) to be able to get cross-attention weights on the fly while the model is ran on segments of audio.
So there is no need to run the model twice (twice faster). It also solves memory issues that could occur on big signals.

Also I think that in the current version of the notebook by @jongwook there is an undesired shift of one token (the cross-attention weights computed on a given input token are relevant for the prediction of the next token).

Please find my code here: https://github.com/Jeronymous/whisper-timestamped
